File: MATLAB/PhenoModel.m
% Phenology Model - NEON Challenge
% x = number of days to simulate
% start = starting date (Jan 1 = day 1)
% G_init = initial proportion of pixels that are green
% a, b, c = rates of movement between compartments

function [G,N] = PhenoModel(x,start,G_init,a,b,c)

% Parameters
beta = @(i) (0 * (i <= 100)) + ...
    (a * (i > 100 && i <= 130)) +...
     (0 * (i > 130 && i <= 290)) +...
     (0 * (i > 290 && i <= 310)) +...
     (0 * (i > 310));
delta = @(i) (0 * (i <= 100)) +...
    (0 * (i > 100 && i <= 130)) +...
    (b * (i > 130 && i <= 290)) +...
    (c * (i > 290 && i <= 310)) +...
    (0 * (i > 310));

% Create Vectors
%G = zeros(start + x + 1); N = zeros(start + x + 1);

% Initial conditions
G(start) = G_init;
N(start) = 1 - G_init;

% Model sim
for i = start+1:(start+x)
    
    G(i) = (1 - delta(i)) * G(i-1) + beta(i) * N(i-1);
    N(i) = (1 - beta(i)) * N(i-1) + delta(i) * G(i-1);

end

end

File: R/PhenoModel.R
# Phenology Model - NEON Challenge

# Input parameters:
  # t = day of year
  # gdd = daily growing degree day value
  # total_days = Accumulated number of growing degree days
  # rolling_avg = 7 day avg GDD 
  # G_init = Initial green value
  # a, b, c, d = growth and death rates (fit)
  # t1, t2, t3, t4 = Timing thresholds (fit)

PhenoModel <- function(GDD,G_init,a,b,c,d,t1,t2,t3,t4) {

  gdd = GDD$GDDdaily
  total_days = GDD$GDDdays
  rolling_avg = GDD$MovAvg_GDDdaily
  n <- length(gdd)
  
  # Parameters
  beta <- a * gdd * (total_days > t1 & total_days <= t2) +          # Green up
    d * ((rolling_avg < t4 & total_days > t3) | total_days <= t1)   # Dormancy
  
  delta <- b * (total_days > t2 & total_days <= t3) +               # Leaf Maturation
    c * (total_days > t3 & rolling_avg > t4)                        # Senescence
  
  # Create Vectors
  G <- rep(0,n)
  N <- rep(0,n)
  
  # Initial conditions
  G[1] = G_init
  N[1] = 1 - G_init
  
  # Model sim
  for ( i in 2:n ) {
    
    G[i] = (1 - delta[i]) * G[i-1] + beta[i] * N[i-1]
    N[i] = (1 - beta[i]) * N[i-1] + delta[i] * G[i-1]
    
  }
  return(G)
  
}
